### PR TITLE
feature: Add PDF support

### PR DESCRIPTION
**Pull request by AhmadMuj**
- Allowed PDF upload in both mobile and web
- Dynamic type is being sent to the server based on the mimetype from both the mobile and web
- Supported PDF view as iframe in both mobile and web
- Added pdf-parse library to parse PDF files
- Added content, info and metadata for the bookmarkAssets table to store the pdf parsing results so they can be indexed in the future. This would also be used in getting a description about the image and storing it in the content field.
- Added inferTagsFromPDF that would parse the PDF, update the bookmarkAssets row and then inferTags from the text that includes the PDF content.

### CODE REVIEW

File: apps/web/app/api/assets/route.ts
Comment (MohamedBassem) [inline apps/web/app/api/assets/route.ts]: For backward compatibility reasons, especially with phones which don't get release immediately, we'll need to keep accepting the `image` field for a while.
```suggestion
  // TODO: Deprecate the `image` field in a couple of releases
  const data = formData.get("file") ?? formData.get("image");
```

File: apps/workers/openaiWorker.ts
Comment (MohamedBassem) [inline apps/workers/openaiWorker.ts]: The difference between this and the text prompt is not that big. I think we can just continue using the text prompt even for pdfs, what do you think?

File: apps/workers/openaiWorker.ts
Comment (MohamedBassem) [inline apps/workers/openaiWorker.ts]: I think this logic fits more into being in the `crawler`. It's the worker type that enriches the `bookmark` in prep for the inference. We should probably rename it as well now that it's no longer about "crawling" :D 

I don't mind keeping this here for the sake of this PR and I can move it after it lands :)

File: apps/workers/openaiWorker.ts
Comment (MohamedBassem) [inline apps/workers/openaiWorker.ts]: nit: This is an enum, let's do a switch maybe?

File: apps/workers/package.json
Comment (MohamedBassem) [inline apps/workers/package.json]: I'm not a big fan of this package. It seems to be abandoned with the last version being from 5 years ago (https://www.npmjs.com/package/pdf-parse?activeTab=versions). 
Should we maybe use https://github.com/mozilla/pdf.js instead? It even comes with a viewer that we can probably use instead of the `iframe`?

File: apps/workers/searchWorker.ts
Comment (MohamedBassem) [inline apps/workers/searchWorker.ts]: I'm a bit surprised that typescript didn't complain about the fields used here as they are not defined in the search schema in `packages/shared/search.ts`

File: packages/db/schema.ts
Comment (MohamedBassem) [inline packages/db/schema.ts]: What exactly gets added in the `info` and `metadata`? Are they really worth storing and indexing?

File: apps/workers/openaiWorker.ts
Comment (MohamedBassem) [inline apps/workers/openaiWorker.ts]: Models have a context window that we need to respect. You'll need to truncate the content to fit in that window especially if someone uploads a large pdf.

There's a function called `truncateContent` that's used in the text prompt builder, maybe you can reuse it?

File: packages/db/schema.ts
Comment (MohamedBassem) [inline packages/db/schema.ts:162]: let's maybe also store the file name?

**Review by MohamedBassem — CHANGES_REQUESTED** [review]
Oh wow. I didn't see this coming!
First of all, thanks a lot for taking the time and contributing with this PR. I'm impressed with how you managed to navigate the different parts of the codebase without any help! As the first contributor to this project, please let me know if there's anything I can improve in the development documentation :)

The PR itself looks good to me, I requested some minor changes. Don't feel obliged to tackle the bigger ones, I'm happy to take over what you won't have time to fix!

I think the main missing thing is mobile rendering for the pdf type asset which lives in `apps/mobile/components/bookmarks/BookmarkCard.tsx`. I think for now, we can start with a static image saying that this is of type pdf and add a download button, later we can add the preview there.

File: apps/web/app/api/assets/route.ts
Comment (ahmadmucom) [reply apps/web/app/api/assets/route.ts]: You're correct. I'll add this

File: apps/workers/openaiWorker.ts
Comment (ahmadmucom) [reply apps/workers/openaiWorker.ts]: You are correct, I'm a big fan of generalization when possible. 
I'll combine them.

File: apps/workers/openaiWorker.ts
Comment (ahmadmucom) [reply apps/workers/openaiWorker.ts]: I would prefer to move it in the next PR to make this one stable.

File: packages/db/schema.ts
Comment (ahmadmucom) [reply packages/db/schema.ts]: In case of a PDF, the metadata will have any PDF meta that are added to the PDF file like the author and so on. 
For the info, it's any relevant information about the PDF version. 

Honestly I kept them for the sake of applying the same for the images as well.
Basically, we are sending the image to openai, why not ask openai to provide both the tags and a description about the image ?
The cost is the same and it will enhance the images search by a huge amount with no extra cost.

File: apps/workers/openaiWorker.ts
Comment (ahmadmucom) [reply apps/workers/openaiWorker.ts]: I had in mind the 4MB limit for the upload. But you are correct I'll add the truncate content.

File: packages/db/schema.ts
Comment (ahmadmucom) [reply packages/db/schema.ts:162]: Well, it's both a negative and a positive thing. 
In case of a "paste" image case, there would be no file name. 
Also many people will upload files with random names. 
I have mixed feelings about this.

File: packages/db/schema.ts
Comment (MohamedBassem) [reply packages/db/schema.ts]: My main concern about them is the fact that they are not structured data and I'm not a big fan of dumping schema-less jsons into the database. Basically, they'll never be useful beyond getting indexed for search. But if you think they provide useful data, then why not. I think we can keep `metadata` and drop `info`? Or merge info into `metadata`?

> why not ask openai to provide both the tags and a description about the image ?

That I 100% agree with and this can be in the `content` field in the database. My comment is in particular about `info` and `metadata`.

File: packages/db/schema.ts
Comment (MohamedBassem) [reply packages/db/schema.ts:162]: It's fine for the `paste`s to not have a filename. My plan is to use the filename of assets to be their `title` in the bookmark grid and also give the users the ability to modify it if they want. I checked and this is the default behavior in raindrop for example. Files without a `filename` would just show no title (like they are now). I think the filename in the pdf case would be a nice way to disambiguate the card without having to look into the content itself.

File: packages/db/schema.ts
Comment (ahmadmucom) [reply packages/db/schema.ts]: I agree.
I'll dump the info and keep the metadata.

File: apps/workers/package.json
Comment (ahmadmucom) [reply apps/workers/package.json]: I think mozilla/pdf.js is not the right tool for us as it's a PDF viewer for HTML 5.
What do you think of this ?
https://github.com/modesty/pdf2json

Comment (MohamedBassem) [reply apps/workers/package.json]: I don't mind pdf2json. It has a weird api but sounds like it can work 👍

Comment (ahmadmucom) [reply apps/workers/package.json]: It has a weird API indeed. I'll work with it for now until we find another good fit.

**Comment by AhmadMuj** [comment]
@MohamedBassem Kindly check now and let me know. 
I'll add the images paste feature later after we close this. 
I changed the prompt slightly to add a description for all content types. It's nice to have a summarization of the content. I didn't support indexing nor handling the content for now as I only worked on the prompt side for now. 
The rest is handled

File: apps/workers/openaiWorker.ts
Comment (MohamedBassem) [inline apps/workers/openaiWorker.ts]: Do you mind if we move the prompt changes to another PR? I'm trying to be very careful with prompt changes as they can have unintended behaviors and they require some testing. For example, here we changed everything to `5-8 tags` and some people were complaining that the `3-5` that I had before were already too much.

I think for the sake of this PR let's just use the TEXT prompt without any changes. It'll probably be good enough :)

File: apps/workers/searchWorker.ts
Comment (MohamedBassem) [inline apps/workers/searchWorker.ts]: There shouldn't be an `info` here? not sure how TS didn't complain as well

File: apps/workers/utils.ts
Comment (MohamedBassem) [inline apps/workers/utils.ts]: We probably should remove those `console.log`s

File: apps/workers/utils.ts
Comment (MohamedBassem) [inline apps/workers/utils.ts]: It seems that `Meta` is `Record<string, any>` according to https://github.com/modesty/pdf2json/blob/master/pdfparser.d.ts#L26 ?

**Review by MohamedBassem — CHANGES_REQUESTED** [review]
Thanks! I think this is good to go after addressing the prompt change comment! I'll take care of adding the mobile renderer in a separate PR.

It also seems that CI is failing, you'll need to run in the root of the repo:

```
pnpm typecheck && pnpm format:fix && pnpm lint:fix
```

to ensure that CI will pass when you push a new version.

File: packages/shared/assetdb.ts
Comment (MohamedBassem) [inline packages/shared/assetdb.ts]: This needs to be nullable for backward compatibility. Otherwise hoarder throws errors when we attempt to read the metadata for assets uploaded before this commit.

```suggestion
  fileName: z.string().nullish(),
```

**Review by MohamedBassem — COMMENTED** [review]
The PR LGTM! We can land after fixing the asset schema and after you restore the preview fix that was lost in the merge commit!

**Pull request merged**
